[VAULT] Add run Vault on zCX page (#1479)
hashicorp/web-unified-docs: content/vault/v1.21.x/content/docs/deploy/run-as-zcx-cluster.mdx

---
layout: docs
page_title: Run Vault on IBM z/OS Container Extensions
description: >-
  Step-by-step guide to deploying a secure 3-node Vault cluster inside IBM zCX using HAProxy and TLS.
---

# Run Vault on IBM z/OS Container Extensions (zCX)

This guide walks through deploying a fully secured 3-node HashiCorp Vault Enterprise cluster
on IBM z/OS Container Extensions (zCX). The setup uses three independent zCX instances, 
each running Vault with its own unique IP address. A Layer-4 HAProxy load balancer sits 
in front to distribute traffic, and end-to-end TLS encryption is applied across all Vault nodes 

> and the load balancer to ensure secure communication throughout the cluster.

Reviewer: ```suggestion
Deploy a fully secured 3-node HashiCorp Vault Enterprise cluster on IBM z/OS
Container Extensions (zCX) with:

- Three independent zCX instances running Vault on unique IP addressed.
- A Layer-4 HAProxy load balancer to distribute traffic.
- End-to-end TLS encryption to ensure secure communication throughout the cluster.
```
Style correction: describe the outcome of following the how-to guide, not the contents of the guide

![Vault zCX cluster deployment](/img/run-as-zcx-cluster.png)

> 

Reviewer: ```suggestion

## Before you start

- **You must have your encryption certificates**. To secure communication in the
  cluster between Vault clients, HAProxy, and other nodes you must have the
  following certificates:
  - **`vault.pem`** - the TLS certificate that secures communication between nodes
  - **`vault.key`** - the private encryption key used by Vault
  - **`ca.pem`** - the certificate authority (CA) used to validate mutual TLS
    between nodes
```

## Step 1: Container image deployment

> 1. Pull the official Vault Enterprise container image on all three zCX nodes.

Reviewer: ```suggestion
1. Use the Docker CLI to pull official Vault Enterprise container images on all three zCX nodes:
```

  ```shell-session
  $ docker pull hashicorp/vault-enterprise
  ```

1. Verify image integrity.

  ```shell-session
  $ docker images | grep vault-enterprise
  ```

> ## Step 2: Persistent volume creation

Reviewer: ```suggestion
## Step 2: Create a persistent volume on each node
```

Create Docker volumes for configuration and data persistence on each node.

1. Create volume for Vault data storage (Raft backend, audit logs).

  ```shell-session
  $ docker volume create vault-data
  ```

> 1. Create volume for Vault configuration files.

Reviewer: ```suggestion
1. Create a separate volume for Vault configuration files:
```

  ```shell-session
  $ docker volume create vault-config
  ```

1. Verify volume creation.

  ```shell-session
  $ docker volume ls | grep vault
  ```

Volume mount points:

- **`vault-data`**: `/vault/data` (stores Raft snapshots, encrypted storage backend)
- **`vault-config`**: `/vault/config.d` (HCL configuration files)

## Step 4: Vault configuration file setup

Create Vault configuration on each node with node-specific parameters.

> 1. Access config volume.

Reviewer: ```suggestion
1. Use the Docker CLI to access the `vault-config` volume:
```

  ```shell-session
  $ docker run --rm -it -v vault-config:/vault alpine sh

>   ```

Reviewer: ```suggestion
```
Since they need to create multiple files and add the certs as well, we typically break the activities into two distinct steps: creating the files and "installing" the files.

1. Create configuration file for Leader and follower vault.

  ```hcl
  ui = true
  disable_mlock = true

  # --- Listener Configuration ---
  listener "tcp" {
    address              = "0.0.0.0:8200"
    tls_disable          = 0
    tls_cert_file        = "/vault/vault.pem"
    tls_key_file         = "/vault/vault.key"
    tls_client_ca_file   = "/vault/ca.pem"
  }

  # --- Raft Storage (Leader Node) ---
  storage "raft" {
    path    = "/vault/data"
    node_id = "<LEADER_NODE_ID>"  # e.g., "node1"

    # No retry_join block needed for leader
  }

  # --- Cluster Networking ---
  api_addr     = "https://<LEADER_IP>:8200"
  cluster_addr = "https://<LEADER_IP>:8201"
  ```

  Exposes Vault on port 8200 for API and UI traffic.

   - **`tls_disable`** = 0 → TLS is enabled (mandatory for secure cluster communication).

### Certificates:

- **`vault.pem`** → TLS certificate

- **`vault.key`** → Private key

- **`ca.pem`** → CA used to validate mutual TLS between nodes

This ensures encrypted traffic between Vault clients, HAProxy, and other nodes.

> Similarly create configuration for follower Vault nodes.

Reviewer: ```suggestion
1. Create a configuration file for Vault follower nodes:. For example, the
   following configuration file exposes Vault on port 8200 for API and UI
   traffic, sets `tls_disable` to `0` to force TLS communication within the
   cluster, and sets the leader information so that followers know how to join
   the cluster:
```

```hcl
ui = true
disable_mlock = true

# --- Listener Configuration ---
listener "tcp" {
  address            = "0.0.0.0:8200"
  tls_disable        = 0
  tls_cert_file      = "/vault/vault.pem"
  tls_key_file       = "/vault/vault.key"
  tls_client_ca_file = "/vault/ca.pem"
}

# --- Raft Storage (Follower Node) ---
storage "raft" {
  path    = "/vault/data"
  node_id = "<FOLLOWER_NODE_ID>"   # e.g., "node2" or "node3"

  retry_join {
    leader_api_addr      = "https://<LEADER_IP>:8200"
    leader_ca_cert_file  = "/vault/ca.pem"
  }
}

# --- Cluster Networking ---
api_addr     = "https://<FOLLOWER_IP>:8200"
cluster_addr = "https://<FOLLOWER_IP>:8201"
```

> 

Reviewer: ````suggestion

## Step 4: Push the configuration files and certificates

1. Use the Docker CLI to access the `vault-config` volume:

  ```shell-session
  $ docker run --rm -it -v vault-config:/vault alpine sh
  ```

1. Add the leader configuration files to the volume:
  
  ```shell-session
  $ TBD
  ```

1. Add the follower configuration files to the volume:
  
  ```shell-session
  $ TBD
  ```

1. Use the Docker CLI to access the `vault-data` volume:

  ```shell-session
  $ docker run --rm -it -v vault-data:/vault alpine sh
  ```

1. 2. Add the certificate files to the volume:  
  ```shell-session
  $ TBD
  ```
````

We need to call out the explicit step for adding the configuration files and the cert files to the relevant nodes

## Step 5: Vault container deployment

> Deploy Vault container on each zCX node.

Reviewer: ```suggestion
Deploy the Vault container on each zCX node.
```

> 1. Set your Vault Enterprise license (export before running).

Reviewer: ```suggestion
1. Set and export the Vault license environment variable, `VAULT_LICENSE` to
   your Enterprise license token.
```

Where are you setting this environment variable? In which terminal? Would it be easier for folks to use the configuration parameter and add a license file to the config mount instead?

  ```shell-session
  $ export VAULT_LICENSE="02MV4UU43BK5HGYYTOJZWFQMTMNNEWU33JJVVGC..." 
  ```

1. Deploying the Vault Enterprise container on zCX.

  ```shell-session
  $ docker run -d \
      --name vault-zcx \
      --cap-add IPC_LOCK \
      -p 8200:8200 \
      -p 8201:8201 \
      -v vault-config:/vault/config.d \
      -v vault-data:/vault/data \
      -e VAULT_LICENSE="$VAULT_LICENSE" \
      hashicorp/vault-enterprise:1.19.12-ent \
      server -config=/vault/config.d/
  ```

> 1. Verify container status.

Reviewer: ```suggestion
1. Check the deployment by verifying the container status:
```

  ```shell-session
  $ docker ps | grep vault-zcx
  ```

>  1. Check the container log. 

Reviewer: ```suggestion
 1. Once Docker reports that the container is running, verify <TBD> by checking the container log:
```
What are they verifying by checking the log? What's the purpose of this step?

  ```shell-session 
  $ docker logs vault-zcx
  ```

## Cluster initialization and unsealing

<Note>

Perform the initialization only on the leader node.

> </Note>

Reviewer: ```suggestion
```
Already called out explicitly in the suggested heading edit

> 1. Initialize the leader Vault.

Reviewer: ```suggestion
1. Use Docker to call the Vault CLI and Initialize the leader node:
```

  ```shell-session
  $ docker exec -it vault-zcx vault operator init \

>     -format=json > /secure/location/vault-init.json

Reviewer 2: This will be path in the container where user want to store unseal key and root token for later use

  ```

>   By default, the `operator init` command generates five unseal keys with threshold of three.

Reviewer: ```suggestion
  By default, `vault operator init` generates five unseal keys and configures
  Vault to require three keys as the unseal threshold.
```

> 1. Extract unseal keys and root token.

Reviewer: ```suggestion
1. Extract unseal keys and root token from `vault-init.json`:
```

  ```shell-session
  $ cat /secure/location/vault-init.json
  ```

1. Unseal Vault on all nodes. 

>   Vault requires a certain threshold of shares to reconstruct the unseal key. Unseal Vault with 3 different unseal keys (repeat on each node).

Reviewer: ```suggestion
```
How-to guides should focus on explaining how to do something, not how the things works. I recommend adding an item to "Before you start" that notes you must be familiar with how unsealing works in Vault.

  ```shell-session
  $ docker exec -it vault-zcx vault operator unseal <UNSEAL_KEY_1>
  $ docker exec -it vault-zcx vault operator unseal <UNSEAL_KEY_2>
  $ docker exec -it vault-zcx vault operator unseal <UNSEAL_KEY_3>
  ```

> 1. Remember to unseal with same key that you got from leader. Check the seal status.

Reviewer: ```suggestion
1. Check the status of Vault:
```

Reviewer 2: To join a Raft peer, the follower vault instance must be unsealed using the same unseal keys generated during the leader’s `vault operator init`. All Raft nodes must share these unseal keys to decrypt the cluster’s storage and communicate securely.

  ```shell-session
  $ vault status
  ``` 

> ## HAProxy load balancer 

Reviewer: ```suggestion
## Step 7: Configure HAProxy load balancer 
```

- Better performance: Forwarding raw TCP reduces overhead, giving lower latency and higher throughput.

### Configuration

```plaintext
global
    maxconn 4096
    log stdout format raw local0

defaults
    mode tcp                     # TCP mode for Layer-4 forwarding
    timeout connect 5s
    timeout client 1m
    timeout server 1m
    log global

# --- Stats Page (HTTPS) ---
frontend stats
    bind *:8404 ssl crt /usr/local/etc/haproxy/haproxy.pem
    mode http
    stats enable
    stats uri /stats
    stats refresh 10s
    stats admin if TRUE

# --- Vault API Frontend (TLS Passthrough) ---
frontend vault_api
    bind *:8200                 # Listen on Vault API port
    mode tcp                    # Layer-4 forwarding to preserve TLS
    default_backend vault_nodes

# --- Vault Nodes Backend ---
backend vault_nodes
    mode tcp
    balance roundrobin          # Simple load distribution among nodes
    option tcp-check            # Health check at TCP level
    server node1 <NODE1_IP>:8200 check
    server node2 <NODE2_IP>:8200 check
    server node3 <NODE3_IP>:8200 check
```

> 1. Copy config to volume.

Author: ☝️ @Shobhit-IBM

  ```shell-session
  $ docker cp haproxy.cfg haproxy-tmp:/usr/local/etc/haproxy/haproxy.cfg
  ```

1. Deploy HAProxy load-balancer.

  ```shell-session
  $ docker run -d \
    --name vault-lb \
    -p 8300:8200 \
    -p 8404:8404 \
    -v haproxy-config:/usr/local/etc/haproxy \
    ibmz-hc-registry.ngrok.dev/haproxy:3.2
  ```

> ## Verify cluster

Reviewer: ```suggestion
## Step 8: Verify your Vault deployment
```

1. List all the leader and follower nodes.

  ```shell-session
  $ vault operator raft list-peers
  ```

> 1. Test container health on browser using HAProxy endpoint.

Reviewer: ```suggestion
1. Test the container health by opening the HAProxy load-balancing endpoint,
   `https://<load_balancer_ip>:<port>/stats` in your browser.
```

  ```plaintext
  https://<LOAD_BALANCER_IP>:<PORT>/stats
  ```

1. Test secure Vault connection and view Raft configuration.

  ```shell-session
  $ curl \
    --cacert <CA_CERT_FILE> \
    --header "X-Vault-Token: <VAULT_TOKEN>" \
    https://<LOAD_BALANCER_IP>:<PORT>/v1/sys/storage/raft/configuration \
    | jq .

>   ```

Reviewer: ````suggestion

<Tabs>

<Tab heading="CLI" group="cli">

  ```shell-session
  $ vault read                      \
    -ca-cert "/path/to/vault.pem"   \
    -format json                    \
    /sys/storage/raft/configuration \
    | jq
  ```

</Tab>

<Tab heading="API" group="api">

  ```shell-session
  $ curl                                                \
    --request POST                                      \
    --header    "X-Vault-Token: ${VAULT_TOKEN}"         \
    --namespace "X-Vault-Namespace: ${VAULT_NAMESPACE}" \
    --cacert <CA_CERT_FILE> \
      ${VAULT_PROXY_ADDR}/v1/sys/storage/raft/configuration \
      | jq .
  ```

</Tab>
</Tabs>

````

We generally want new content to provide examples using both the CLI and the API so folks have example code regardless of which method they prefer

## Additional resources

- [Vault configuration parameters](/vault/docs/configuration)

> - [Seal/unseal](/vault/docs/concepts/seal)

Reviewer: ```suggestion
```
Folded into "Before you start"

   API and UI traffic, sets `tls_disable` to `0` to force TLS communication
   within the cluster, and sets the leader information so that followers know
   how to join the cluster:

```hcl
ui = true
disable_mlock = true

# --- Listener Configuration ---
listener "tcp" {
  address            = "0.0.0.0:8200"
  tls_disable        = 0
  tls_cert_file      = "/vault/vault.pem"
  tls_key_file       = "/vault/vault.key"
  tls_client_ca_file = "/vault/ca.pem"
}

# --- Raft Storage (Follower Node) ---
storage "raft" {
  path    = "/vault/data"
  node_id = "<FOLLOWER_NODE_ID>"   # e.g., "node2" or "node3"

  retry_join {
    leader_api_addr      = "https://<LEADER_IP>:8200"
    leader_ca_cert_file  = "/vault/ca.pem"
  }
}

# --- Cluster Networking ---
api_addr     = "https://<FOLLOWER_IP>:8200"
cluster_addr = "https://<FOLLOWER_IP>:8201"
```

## Step 4: Push the configuration files and certificates

1. Use the Docker CLI to access the `vault-config` volume:

  ```shell-session
  $ docker run --rm -it -v vault-config:/vault alpine sh
  ```

1. Add the leader configuration files to the volume:
  
  ```shell-session

>   $ TBD

Author: ☝️ @Shobhit-IBM What command should go here?

  ```

1. Add the follower configuration files to the volume:
  
  ```shell-session

>   $ TBD

Reviewer: Missing actual command

Reviewer 2: If user have all these file say tls certificate and vault.hcl say in folder : vault-config/

docker run --name vault-tmp -v vault-config:/vault-config alpine sh -c "sleep 1" \ && docker cp /home/admin/vault-config/. vault-tmp:/vault-config/ \ && docker rm -f vault-tmp

he can use this command to copy respective config file and repeat it on all 3 nodes

  ```

1. Use the Docker CLI to access the `vault-data` volume:

  ```shell-session
  $ docker run --rm -it -v vault-data:/vault alpine sh
  ```

1. 2. Add the certificate files to the volume:  
  ```shell-session

>   $ TBD

Reviewer 2: To copy file from host env to docker user can use this command : `docker run --name vault-tmp -v vault-config:/vault-config alpine sh -c "sleep 1" \ && docker cp /home/admin/vault-config/. vault-tmp:/vault-config/ \ && docker rm -f vault-tmp`
